### **1. Introduction to Deep Learning with PyTorch**

#### **1.1 Introduction to PyTorch, a Deep Learning Library**

In [94]:
import torch
import numpy as np

**Tensor from a List**

In [95]:
lst = [[1,2,3], [4,5,6] , [7,8,9]]
tensor = torch.tensor(lst)
print(tensor)

tensor([[1, 2, 3],
        [4, 5, 6],
        [7, 8, 9]])


**Tensor from a NumPy Array**

In [96]:
array = [[1,2,3], [4,5,6] ]
np_array = np.array(array)
np_tensor = torch.from_numpy(np_array)
print(np_tensor)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


**Tensor Attributes**

In [97]:
print(tensor.shape)
print(np_tensor.shape)

torch.Size([3, 3])
torch.Size([2, 3])


In [98]:
print(tensor.dtype)
print(np_tensor.dtype)

torch.int64
torch.int32


In [99]:
tensor.device

device(type='cpu')

**Tensor operations**

In [100]:
# Addition
a = torch.tensor([[1,1],[2,2]])
b = torch.tensor([[2,2],[3,3]])
print(a+b)

tensor([[3, 3],
        [5, 5]])


In [101]:
# Multiplication
a = torch.tensor([[1,1],[2,2]])
b = torch.tensor([[2,2],[3,3]])
print(a*b)

tensor([[2, 2],
        [6, 6]])


##### **First neural Network**

In [102]:
import torch.nn as nn

input_tensor = torch.Tensor([[2, 3, 6, 7, 9, 3, 2, 1]])

# Implement a small neural network with exactly two linear layers
model = nn.Sequential(nn.Linear(8, 4),  #1st Linear Layer
                      nn.Linear(4, 1),  #2nd Linear Layer
                     )

output = model(input_tensor)
print(output)

tensor([[-2.5365]], grad_fn=<AddmmBackward0>)


In [103]:
input_tensor = torch.Tensor([[2, 3, 6, 7, 9, 3, 2, 1]])

# Implement a small neural network with exactly two linear layers
model = nn.Sequential(nn.Linear(8, 4),  #1st Linear Layer
                      nn.Linear(4, 1),  #2nd Linear Layer
                      nn.Sigmoid()
                     )

output = model(input_tensor)
print(output)

tensor([[0.1462]], grad_fn=<SigmoidBackward0>)


**Create a sigmoid function and apply it on input_tensor**

In [104]:
input_tensor = torch.tensor([[0.8]])
sigmoid = nn.Sigmoid()
probability = sigmoid(input_tensor)
print(probability)

tensor([[0.6900]])


**Create a softmax function and apply it on input_tensor**

In [105]:
input_tensor = torch.tensor([[1.0, -6.0, 2.5, -0.3, 1.2, 0.8]])
softmax = nn.Softmax(dim=-1)
probabilities = softmax(input_tensor)
print(probabilities)

tensor([[1.2828e-01, 1.1698e-04, 5.7492e-01, 3.4961e-02, 1.5669e-01, 1.0503e-01]])


#### **1.2 Training Our First Neural Network with PyTorch**

**Recap**

**(i) Binary Classification (foward pass)**

In [106]:
input_tensor = torch.Tensor([[3, 4, 6, 2, 3, 6, 8, 9]])

# Implement a small neural network for binary classification
model = nn.Sequential(
  nn.Linear(8, 1),
  nn.Sigmoid()
)

output = model(input_tensor)
print(output)


tensor([[0.1759]], grad_fn=<SigmoidBackward0>)


**(ii) Multiclass Classification (foward pass)**

In [107]:
input_tensor = torch.Tensor([[3, 4, 6, 7, 10, 12, 2, 3, 6, 8, 9]])

n_classes = 3 
# Update network below to perform a multi-class classification with four labels
model = nn.Sequential(
  nn.Linear(11, 20),
  nn.Linear(20, 12),
  nn.Linear(12, 6),
  nn.Linear(6,n_classes), 
  nn.Softmax(dim = -1)
)

output = model(input_tensor)
print(output)

tensor([[0.4247, 0.3279, 0.2473]], grad_fn=<SoftmaxBackward0>)


**(iii) Regression (foward pass)**

In [108]:
input_tensor = torch.Tensor([[3, 4, 6, 7, 10, 12, 2, 3, 6, 8, 9]])

# Implement a neural network with exactly four linear layers
model = nn.Sequential(
  nn.Linear(11, 20),
  nn.Linear(20, 12),
  nn.Linear(12, 6),
  nn.Linear(6, 1),  
)

output = model(input_tensor)
print(output)

tensor([[-0.3449]], grad_fn=<AddmmBackward0>)


##### **Using loss functions to assess model predictions**

**Creating one-hot encoded labels**

In [109]:
import numpy as np
import torch.nn.functional as F


y = 1
num_classes = 3

# Create the one-hot encoded vector using NumPy
one_hot_numpy = np.array([0, 1, 0])

# Create the one-hot encoded vector using PyTorch
one_hot_pytorch = F.one_hot(torch.tensor(y), num_classes)

print(one_hot_numpy)
print(one_hot_pytorch)

[0 1 0]
tensor([0, 1, 0])


**Calculating entropy loss**

In [110]:
y = [2]
scores = torch.tensor([[0.1, 6.0, -2.0, 3.2]])

# Create a one-hot encoded vector of the label y
one_hot_label = F.one_hot(torch.tensor(y), scores.shape[1])

# Create the cross entropy loss function
criterion = nn.CrossEntropyLoss()

# Calculate the cross entropy loss
loss = criterion(scores.double(), one_hot_label.double())
print(loss)

tensor(8.0619, dtype=torch.float64)


#### **1.3 Neural Network Architecture and Hyperparameters**

#### **1.4 Evaluating and Improving Models**